File: server-rs/src/workspace/worktrees.rs
use crate::db::Db;
use crate::error::{project_error, worktree_error, AppError, AppResult};
use crate::git::{
    create_worktree, list_worktrees as git_list_worktrees, worktree_path_exists, CreateWorktreeInput,
    GitError,
};
use crate::types::{now_iso, Worktree};
use rusqlite::params;

use super::projects::get_project;

fn ms_to_iso(ms: i64) -> String {
    chrono::DateTime::from_timestamp_millis(ms)
        .map(|dt| dt.to_rfc3339_opts(chrono::SecondsFormat::Millis, true))
        .unwrap_or_else(now_iso)
}

fn now_ms() -> i64 {
    chrono::Utc::now().timestamp_millis()
}

fn row_to_worktree(
    id: String,
    project_id: String,
    path: String,
    branch: Option<String>,
    base_branch: Option<String>,
    git_dir: Option<String>,
    is_linked: i64,
    created_at: i64,
) -> Worktree {
    Worktree {
        id,
        project_id,
        path,
        branch,
        base_branch,
        git_dir,
        is_linked: is_linked != 0,
        created_at: ms_to_iso(created_at),
    }
}

fn fetch_worktree(db: &Db, id: &str) -> AppResult<Option<Worktree>> {
    let conn = db.conn();
    let mut stmt = conn
        .prepare(
            "SELECT id, project_id, path, branch, base_branch, git_dir, is_linked, created_at
             FROM worktrees WHERE id = ?1 LIMIT 1",
        )
        .map_err(|e| AppError::Internal(e.to_string()))?;

    let result = stmt.query_row(params![id], |row| {
        Ok(row_to_worktree(
            row.get(0)?,
            row.get(1)?,
            row.get(2)?,
            row.get(3)?,
            row.get(4)?,
            row.get(5)?,
            row.get(6)?,
            row.get(7)?,
        ))
    });

    match result {
        Ok(wt) => Ok(Some(wt)),
        Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
        Err(e) => Err(AppError::Internal(e.to_string())),
    }
}

pub fn get_worktree(db: &Db, id: &str) -> AppResult<Option<Worktree>> {
    fetch_worktree(db, id)
}

pub fn sync_worktrees_for_project(db: &Db, project_id: &str) -> AppResult<()> {
    let project = get_project(db, project_id)?
        .ok_or_else(|| project_error("Project not found", 404))?;

    let git_entries = git_list_worktrees(&project.repo_path)
        .map_err(|e| AppError::Internal(e.message))?;

    let conn = db.conn();
    let mut stmt = conn
        .prepare(
            "SELECT id, project_id, path, branch, base_branch, git_dir, is_linked, created_at
             FROM worktrees WHERE project_id = ?1",
        )
        .map_err(|e| AppError::Internal(e.to_string()))?;

    let existing = stmt
        .query_map(params![project_id], |row| {
            Ok(row_to_worktree(
                row.get(0)?,
                row.get(1)?,
                row.get(2)?,
                row.get(3)?,
                row.get(4)?,
                row.get(5)?,
                row.get(6)?,
                row.get(7)?,
            ))
        })
        .map_err(|e| AppError::Internal(e.to_string()))?
        .collect::<Result<Vec<Worktree>, _>>()
        .map_err(|e| AppError::Internal(e.to_string()))?;

    let mut by_path: std::collections::HashMap<String, Worktree> = existing
        .into_iter()
        .map(|wt| (wt.path.clone(), wt))
        .collect();

    for entry in git_entries {
        let linked = worktree_path_exists(&entry.path);
        if let Some(prev) = by_path.remove(&entry.path) {
            let branch = entry.branch.or(prev.branch);
            let git_dir = entry.git_dir.or(prev.git_dir);
            conn.execute(
                "UPDATE worktrees SET branch = ?1, git_dir = ?2, is_linked = ?3 WHERE id = ?4",
                params![branch, git_dir, linked as i64, prev.id],
            )
            .map_err(|e| AppError::Internal(e.to_string()))?;
        } else {
            let id = uuid::Uuid::new_v4().to_string();
            let created_at = now_ms();
            conn.execute(
                "INSERT INTO worktrees (id, project_id, path, branch, base_branch, git_dir, is_linked, created_at)
                 VALUES (?1, ?2, ?3, ?4, NULL, ?5, ?6, ?7)",
                params![
                    id,
                    project_id,
                    entry.path,
                    entry.branch,
                    entry.git_dir,
                    linked as i64,
                    created_at,
                ],
            )
            .map_err(|e| AppError::Internal(e.to_string()))?;
        }
    }

    for orphan in by_path.values() {
        conn.execute(
            "UPDATE worktrees SET is_linked = 0 WHERE id = ?1",
            params![orphan.id],
        )
        .map_err(|e| AppError::Internal(e.to_string()))?;
    }

    Ok(())
}

pub fn list_worktrees_by_project(db: &Db, project_id: &str) -> AppResult<Vec<Worktree>> {
    if get_project(db, project_id)?.is_none() {
        return Err(project_error("Project not found", 404));
    }
    sync_worktrees_for_project(db, project_id)?;

    let conn = db.conn();
    let mut stmt = conn
        .prepare(
            "SELECT id, project_id, path, branch, base_branch, git_dir, is_linked, created_at
             FROM worktrees WHERE project_id = ?1",
        )
        .map_err(|e| AppError::Internal(e.to_string()))?;

    let rows = stmt
        .query_map(params![project_id], |row| {
            Ok(row_to_worktree(
                row.get(0)?,
                row.get(1)?,
                row.get(2)?,
                row.get(3)?,
                row.get(4)?,
                row.get(5)?,
                row.get(6)?,
                row.get(7)?,
            ))
        })
        .map_err(|e| AppError::Internal(e.to_string()))?;

    rows.collect::<Result<Vec<_>, _>>()
        .map_err(|e| AppError::Internal(e.to_string()))
}

pub fn create_worktree_for_project(
    db: &Db,
    project_id: &str,
    branch: &str,
    base_branch: Option<&str>,
    path: Option<&str>,
    is_new_branch: Option<bool>,
) -> AppResult<Worktree> {
    let project = get_project(db, project_id)?
        .ok_or_else(|| project_error("Project not found", 404))?;

    let result = create_worktree(&CreateWorktreeInput {
        repo_path: &project.repo_path,
        branch: branch.trim(),
        base_branch: base_branch.map(str::trim),
        path: path.map(str::trim),
        is_new_branch,
    })
    .map_err(|e: GitError| worktree_error(e.message, 400))?;

    sync_worktrees_for_project(db, project_id)?;

    let conn = db.conn();
    let mut stmt = conn
        .prepare(
            "SELECT id, project_id, path, branch, base_branch, git_dir, is_linked, created_at
             FROM worktrees WHERE project_id = ?1",
        )
        .map_err(|e| AppError::Internal(e.to_string()))?;

    let rows = stmt
        .query_map(params![project_id], |row| {
            Ok(row_to_worktree(
                row.get(0)?,
                row.get(1)?,
                row.get(2)?,
                row.get(3)?,
                row.get(4)?,
                row.get(5)?,
                row.get(6)?,
                row.get(7)?,
            ))
        })
        .map_err(|e| AppError::Internal(e.to_string()))?;

    let worktrees: Vec<Worktree> = rows
        .collect::<Result<Vec<_>, _>>()
        .map_err(|e| AppError::Internal(e.to_string()))?;

    let matched = worktrees
        .into_iter()
        .find(|w| w.path == result.path)
        .ok_or_else(|| worktree_error("Worktree created but not found after sync", 400))?;

    if let Some(ref base) = result.base_branch {
        conn.execute(
            "UPDATE worktrees SET base_branch = ?1 WHERE id = ?2",
            params![base, matched.id],
        )
        .map_err(|e| AppError::Internal(e.to_string()))?;
    }

    fetch_worktree(db, &matched.id)?
        .ok_or_else(|| worktree_error("Worktree not found", 404))
}

pub fn delete_worktree(db: &Db, id: &str) -> AppResult<()> {
    if fetch_worktree(db, id)?.is_none() {
        return Err(worktree_error("Worktree not found", 404));
    }
    db.conn()
        .execute("DELETE FROM worktrees WHERE id = ?1", params![id])
        .map_err(|e| AppError::Internal(e.to_string()))?;
    Ok(())
}

File: server/modules/workspace/worktrees.ts
import { randomUUID } from "node:crypto";
import { resolve } from "node:path";
import { eq } from "drizzle-orm";
import type { AppDatabase } from "../../db/index.js";
import { worktrees } from "../../db/schema.js";
import { GitError } from "../git/exec.js";
import { createWorktree } from "../git/worktree-create.js";
import {
  listWorktrees as gitListWorktrees,
  worktreePathExists,
} from "../git/worktree-list.js";
import { removeWorktree as gitRemoveWorktree } from "../git/worktree-remove.js";
import { getProject, ProjectError } from "./projects.js";

export class WorktreeError extends Error {
  constructor(
    message: string,
    readonly status: 400 | 404 = 400,
  ) {
    super(message);
    this.name = "WorktreeError";
  }
}

export async function getWorktree(db: AppDatabase["db"], id: string) {
  const rows = await db.select().from(worktrees).where(eq(worktrees.id, id)).limit(1);
  return rows[0] ?? null;
}

export async function syncWorktreesForProject(db: AppDatabase["db"], projectId: string) {
  const project = await getProject(db, projectId);
  if (!project) throw new ProjectError("Project not found", 404);

  const gitEntries = gitListWorktrees(project.repoPath);

  const existing = await db
    .select()
    .from(worktrees)
    .where(eq(worktrees.projectId, projectId));

  const byPath = new Map(existing.map((w) => [w.path, w]));

  for (const entry of gitEntries) {
    const linked = worktreePathExists(entry.path);
    const prev = byPath.get(entry.path);
    if (prev) {
      await db
        .update(worktrees)
        .set({
          branch: entry.branch ?? prev.branch,
          gitDir: entry.gitDir ?? prev.gitDir,
          isLinked: linked,
        })
        .where(eq(worktrees.id, prev.id));
      byPath.delete(entry.path);
    } else {
      await db.insert(worktrees).values({
        id: randomUUID(),
        projectId,
        path: entry.path,
        branch: entry.branch ?? null,
        baseBranch: null,
        gitDir: entry.gitDir ?? null,
        isLinked: linked,
        createdAt: new Date(),
      });
    }
  }

  for (const orphan of byPath.values()) {
    await db
      .update(worktrees)
      .set({ isLinked: false })
      .where(eq(worktrees.id, orphan.id));
  }
}

export async function listWorktreesByProject(db: AppDatabase["db"], projectId: string) {
  const project = await getProject(db, projectId);
  if (!project) throw new ProjectError("Project not found", 404);
  await syncWorktreesForProject(db, projectId);
  return db.select().from(worktrees).where(eq(worktrees.projectId, projectId));
}

export interface CreateWorktreeBody {
  branch: string;
  baseBranch?: string;
  path?: string;
  isNewBranch?: boolean;
}

export async function createWorktreeForProject(
  db: AppDatabase["db"],
  projectId: string,
  body: CreateWorktreeBody,
) {
  const project = await getProject(db, projectId);
  if (!project) throw new ProjectError("Project not found", 404);

  try {
    const result = createWorktree({
      repoPath: project.repoPath,
      branch: body.branch.trim(),
      baseBranch: body.baseBranch?.trim(),
      path: body.path?.trim(),
      isNewBranch: body.isNewBranch,
    });

    await syncWorktreesForProject(db, projectId);
    const rows = await db
      .select()
      .from(worktrees)
      .where(eq(worktrees.projectId, projectId));
    const match = rows.find((w) => w.path === result.path);
    if (!match) {
      throw new WorktreeError("Worktree created but not found after sync");
    }

    if (result.baseBranch) {
      await db
        .update(worktrees)
        .set({ baseBranch: result.baseBranch })
        .where(eq(worktrees.id, match.id));
    }

    return (await getWorktree(db, match.id))!;
  } catch (err) {
    if (err instanceof GitError) {
      throw new WorktreeError(err.message);
    }
    throw err;
  }
}

export async function deleteWorktree(db: AppDatabase["db"], id: string) {
  const row = await getWorktree(db, id);
  if (!row) throw new WorktreeError("Worktree not found", 404);

  const project = await getProject(db, row.projectId);
  if (!project) throw new WorktreeError("Project not found", 404);

  if (resolve(project.repoPath) === resolve(row.path)) {
    throw new WorktreeError("Cannot remove the main worktree", 400);
  }

  if (worktreePathExists(row.path)) {
    try {
      gitRemoveWorktree(project.repoPath, row.path, { force: true });
    } catch (err) {
      if (err instanceof GitError) {
        throw new WorktreeError(err.message);
      }
      throw err;
    }
  }

  await db.delete(worktrees).where(eq(worktrees.id, id));
}
